src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs

```diff
@@ -534,6 +546,16 @@ public virtual bool CanDeserialize(XmlReader xmlReader)
             {
                 return _tempAssembly.CanRead(_mapping, xmlReader);
             }
+            else if (ShouldUseReflectionBasedSerialization(_mapping) || _isReflectionBasedSerializer)
```

Comment:
@HongGit @StephenMolloy - one `XmlSerializer\ReflectionOnly` test was failing with the other changes because `CanDeserialize` was returning `false` when `_primitiveType == null` and `_tempAssembly == null`. So I wrote this logic to handle that case. Can you take a look?

I plan on porting this PR back to 6.0, so it would be good to get eyes on it.

Reply:
@MichalStrehovsky - any thoughts on why that `XmlSerializer\ReflectionOnly` test doesn't fail in runtimelab's NativeAOT branch?

Reply:
We don't run those tests in the NativeAOT branch. The set of fixes we have in the NativeAOT branch comes from manually porting the UWP ifdefs from an older version of the source code.

Looks like I missed this path.

This is how it looked like:

https://github.com/dotnet/corefx/blob/e31a2b7f3bdf985f56798df530daa4c06fd80376/src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs#L682-L700

Reply:
Thanks, @MichalStrehovsky. I took the UAP logic (using `ReflectionMethodEnabled`) and brought it into this PR.

Reply:
> so I would move this condition ahead in the pecking order

I can switch the `if` conditions around so they match `Deserialize`. That makes sense to me. Originally I did it this way "just in case" it doesn't break someone when `_tempAssembly` _is_ available. But it should logically match `Deserialize`.

> I would also note that when matching this condition, we will try to do reflection-based serialization, period. No fallback. So I might consider just returning true, since returning ReflectionMethodEnabled

Honestly - the `Mode` property in this code is confusing/convoluted. Does anyone actually set it to `CodeGenOnly` or `PreGenOnly`? I don't see any code (other than tests) doing that.

Since `Mode` can never be anything other than `ReflectionOnly` or `ReflectionAsBackup`, I can just return `true` here. But if we did want to respect `Mode` for `CodeGenOnly` or `PreGenOnly`, I think it should logically use `ReflectionMethodEnabled`. You are right that if you had a SOAP mapping, and call `Deserialize` with `Mode == CodeGenOnly | PreGenOnly`, nothing is checking the Mode, so it should work.

Reply:
Well... um... yeah. Like I said, XmlSerializer has been cobbled together from many different iterations on previous platforms. So it's kind of a mess. Reflection-only is a .Net Core thing I believe - just kind of stapled on to the mess that was already XmlSerializer moving through various runtimes. As we've hinted at previously, we do hope to clean this all up in 7.0 as we try to move to reflection-based serialization as the primary mode of operation rather than a niche alternative.